### ISSUE TITLE
Union of tuples not narrowed as expected

### ISSUE THREAD

**Issue by xylix** [microsoft/pyright#2384]
**Describe the bug**
Pyright not narrowing down certain unions of two tuples.

**To Reproduce**

Type check the code sample below, tested with pyright command line pyright 1.1.175.

**Expected behavior**
Pyright to narrow down  the type after asserting a field of the tuple.

**Screenshots or Code**
 
```python
 
import random
from typing import TypeVar, Tuple, Union, cast

T = TypeVar("T")

Maybe = Union[Tuple[T, None], Tuple[None, str]]


def test():
    if random.choice([True, False]):
        return None, "Turd"
    return 1, None


def works(a: Maybe[T]) -> Tuple[T, None]:
    assert a[1] is None
    # Need manual cast
    a = cast(Tuple[T, None], a)
    return a


def fails(a: Maybe[T]) -> Tuple[T, None]:
    assert a[1] is None

    return a # Inferred type is Maybe[T] instead of the left branch of Maybe
 
```

**VS Code extension or command-line**
Command line, version `pyright 1.1.175`

**Additional context**

**Comment by erictraut** [comment]
This functionality will be included in the next release of pyright.
